csharp/src/Apache.Arrow/Ipc/ArrowFileReaderImplementation.cs

```diff
@@ -0,0 +1,195 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using FlatBuffers;
+using System;
+using System.Buffers.Binary;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Apache.Arrow.Ipc
+{
+    internal class ArrowFileReaderImplementation : ArrowStreamReaderImplementation
+    {
+        public bool IsFileValid { get; private set; }
+
+        /// <summary>
+        /// When using GetNextRecordBatch this value 
+        /// is to remember what index is next
+        /// </summary>
+        private int _recordBatchIndex;
+
+        /// <summary>
+        /// Notes what byte position where the footer data is in the stream
+        /// </summary>
+        private int _footerStartPostion;
+
+        private ArrowFooter _footer;
+
+        public ArrowFileReaderImplementation(Stream stream) : base(stream)
+        {
+        }
+
+        public async Task<int> RecordBatchCountAsync()
+        {
+            if (!HasReadSchema)
+            {
+                await ReadSchemaAsync();
```

Comment:
As this is a library that could be used in a wide variety of contexts, we should use `.ConfigureAwait(false)` on all awaits.

Reply:
Done.